### ISSUE TITLE
clipboard between freerdp/freerdp

### ISSUE THREAD

**Issue by FragsterAt** [FreeRDP/FreeRDP#3461]
clipboard corrupted, when copy from one freerdp session and try to paste in other session.
freerdp-2.0.0~git.1463131968.4e66df7 (precompiled for opensuse 42.1)

**Comment by ilammy** [comment]
Hi @FragsterAt,

Could you please elaborate on the use case you are experiencing the issue with?

- What applications do you use in both sessions?
- What kind of data are you trying to copy and paste?
- Are you able to reproduce the corruption when using an intermediate local application as a buffer?

Also, could you please describe the nature of the corruptions?

- Do you see garbage instead of the expected data, or just some minor changes?
- Is the data truncated, or some part of it missing?

Thank you.

I could not reproduce the issue with both latest master (5b2455f3147aca395a441fc65b602938600f0788) and the specified version (4e66df7228b14e432a0793d32ac2524d37cb88ec). I used Mozilla Firefox, MS Office apps, Paint, Notepad, Wordpad. All of them have been able to exchange text, HTML, and images just fine.

The case when the data is transferred directly between two remote sessions is a special one in FreeRDP. In this case FreeRDP does not perform any conversions and only acts an intermediate party between two remote clipboards, simply transferring blobs of data.

The only expected corruption would be truncation of the raw clipboard data when it does not fit into X11 property. This could be fixed, I suppose, by implementing the INCR mechanism for huge amounts of data. 

However, any other kind of corruption would be unexpected in this case and is an indicator of something strange happening.

**Comment by FragsterAt** [comment]
Copy-paste from notepad to notepad. In host system paste into kate (it works).

![screenshot_freerdp](https://cloud.githubusercontent.com/assets/20900647/17581970/1d05d98e-5fba-11e6-802f-2ecc01558f1d.png)

If paste in kate, then copy from kate and paste in 2nd session - it also ok.

**Comment by meersjo** [comment]
@FragsterAt , can you verify contents of all three clipboards (primary, secondary and clipboard) using xsel or similar utilities? Is it possible that the chinese (?) text was taken from the wrong clipboard? Depending on how you copy in X, it may set multiple clipboards to the same value.

**Comment by ilammy** [comment]
@meersjo, I do not think that is the case. FreeRDP uses only CLIPBOARD, so other buffers should not interfere.

I believe this may have to do something with different locales or with remote apps being used. Like, CF_OEMTEXT from Russian locale being incorrectly converted into CF_UNICODETEXT on the English side. Here is the source string in some popular encodings:
```
 H  e  l  l  o \n  П  р  и  в  е  т
48 65 6C 6C 6F 0A CF F0 E8 E2 E5 F2  Windows-1251
48 65 6C 6C 6F 0A 8F E0 A8 A2 A5 E2  CP866
48 65 6C 6C 6F 0A F0 D2 C9 D7 C5 D4  KOI-8R
```
And here is destination string interpreted as UTF-16LE:
```
6548 6C6C 0A6F F0CF E2E8 F2E5  Windows-1251
效汬੯
6548 6C6C 0A6F E08F A2A8 E2A5  CP866
效汬੯ꊨ
6548 6C6C 0A6F D2F0 D7C9 D4C5  KOI-8R
效汬੯티퟉퓅
```
I tried to be more creative with byte permutations, but after spending two hours doing this and searching for these CJK glyphs I got bored. At least the first glyphs look alike to what is seen in the screenshot.

However, I have tried using English and Russian locales, but notepads worked fine with whatever encoding I had selected. Unfortunately, the environment I have at the moment does not support remote apps, so I am not able to reproduce the use case precisely now. I should be able to do this on workweek.

Meanwhile, @FragsterAt, it would help if you could log whatever is happening with the clipboard channel in your environment. Unfortunately, the release build is lacking in debug logs, so you will have to rebuild FreeRDP to see anything useful. Here are the steps

- Install [build prerequisites](https://github.com/FreeRDP/FreeRDP/wiki/Compilation)

- Grab your version from Git:

   ```
   $ cd /tmp
   $ git clone https://github.com/FreeRDP/FreeRDP.git
   $ cd FreeRDP
   $ git checkout 4e66df7228b14e432a0793d32ac2524d37cb88ec
   ```

- Configure and build with logging enabled:

   ```
   $ cmake -DCMAKE_BUILD_TYPE=Debug -DWITH_DEBUG_CLIPRDR=ON .
   $ make -j4
   ```

- Enable FreeRDP logging for clipboard and run your clients like this:

   ```
   $ export WLOG_FILTER=com.freerdp.channels.cliprdr.client:DEBUG
   $ ./client/X11/xfreerdp /u:user /v:host1 +clipboard /app:"notepad" | tee host1.log &
   $ ./client/X11/xfreerdp /u:user /v:host2 +clipboard /app:"notepad" | tee host2.log &
   ```

- If everything works fine you should see lines like this when copying and pasting:

   ```
[01:46:27:100] [27816:27822] [DEBUG][com.freerdp.channels.cliprdr.client] - msgType: CB_FORMAT_DATA_RESPONSE (5), msgFlags: 1 dataLen: 28
[01:46:27:100] [27816:27822] [DEBUG][com.freerdp.channels.cliprdr.client] - 0000 05 00 01 00 1c 00 00 00 48 00 65 00 6c 00 6c 00 ........H.e.l.l.
[01:46:27:100] [27816:27822] [DEBUG][com.freerdp.channels.cliprdr.client] - 0010 6f 00 0d 00 0a 00 1f 04 40 04 38 04 32 04 35 04 o.......@.8.2.5.
[01:46:27:100] [27816:27822] [DEBUG][com.freerdp.channels.cliprdr.client] - 0020 42 04 00 00                                     B...
[01:46:27:100] [27816:27822] [DEBUG][com.freerdp.channels.cliprdr.client] - ServerFormatDataResponse
   ```

Please launch both notepads like this (using different files), reproduce the issue, and drop the logs here. This would be a great help as it should show what exactly goes wrong with the conversion, which encodings are used, etc. Meanwhile I will examine the conversion code for any possible issues that may be responsible for such behavior.

**Comment by ilammy** [comment]
Ugh, actually, I just got an insight immediately after giving the bytes another look and noticing that there are more glyphs than half of the source bytes.

```
 H  e  l  l  o \n   П     р     и     в     е     т     Source string 
48 65 6C 6C 6F 0A D0 9F D1 80 D0 B8 D0 B2 D0 B5 D1 82   Source string encoded in UTF-8
6548  6C6C  0A6F  9FD0  80D1  B8D0  B2D0  B5D0  82D1    Interpret UTF-8 bytes as UTF-16LE
 效    汬     ੯     鿐    胑    룐    닐     뗐    苑     Destination string
```

Well, there is one extra glyph in the end, but otherwise this seems like the situation we have here.

Now it's interesting to see what formats can cause this, as I do not see right away any fallthrough branches in FreeRDP conversions which work with Unicode and do not call `Convert{To,From}Unicode()`. However, LF line ending is suspicious: Windows would have used CR-LF.

**Comment by FragsterAt** [comment]
build Freerdp from git 4e66df7228b14e432a0793d32ac2524d37cb88ec

command line was like ~/FreeRDP/client/X11/xfreerdp /u:<usr> /p:<pwd> /d:<domain> +clipboard /v: | tee <log> in both sessions (connect in 2 different servers, 1st win10, 2nd win server 2012)

| tee <log> part didn't work so I attach console output
1- copy from srv1, paste in host and srv2
2 - copy pasted text from host and paste in srv1 and srv2


[srv1.txt](https://github.com/FreeRDP/FreeRDP/files/417733/srv1.txt)
[srv2.txt](https://github.com/FreeRDP/FreeRDP/files/417734/srv2.txt)
![screenshot_20160815_111244](https://cloud.githubusercontent.com/assets/20900647/17659316/b88a36fe-62e2-11e6-8c6e-cf4ec474a746.png)

**Comment by ilammy** [comment]
**tl;dr:** I know the cause of the issue. @FragsterAt, I will ping you when a fix is ready.

----

O-o-okay. So here is a transcription of the logs.

```
[11:10:56:585]  Desktop 1 -> Client   My available clipboard formats:
                                        0x000D CF_UNICODETEXT
                                        0x0010 CF_LOCALE
                                        0x0001 CF_TEXT
                                        0x0007 CF_OEMTEXT

[11:10:56:585]  Client -> Desktop 1   Thanks.

[11:10:56:586]  Client -> Desktop 2   My available clipboard formats:
                                        0x0000 CF_RAW
                                        0x000D CF_UNICODETEXT
                                        0x0001 CF_TEXT
                                        0xD011 ?
                                        0xD012 ?
                                        0xD013 ?
                                        0x0008 CF_DIB
                                        0xD010 "HTML Format"

[11:10:56:587]  Client -> Desktop 2   My available clipboard formats:
                                        0x000D CF_UNICODETEXT
                                        0x0010 CF_LOCALE
                                        0x0001 CF_TEXT
                                        0x0007 CF_OEMTEXT

[11:10:56:638]  Client -> Desktop 1   Give me CF_UNICODETEXT, please.

[11:10:56:642]  Desktop 1 -> Client   Here you go:
                                        "Hello\r\nПривет\0" encoded in UTF-16LE
                                        with terminating null byte, with a line
                                        ending as CR-LF, exactly as in the spec

[11:10:56:643]  Desktop 2 -> Client   Thanks. Thanks.

[11:11:04:955]  Desktop 2 -> Client   Give me CF_UNICODETEXT, please.

[11:11:04:956]  Client -> Desktop 2   Here you go:
                                        "Hello\nПривет" encoded in UTF-8,
                                        without a terminating null byte,
                                        with a LF line ending
```

Important points:

- source desktop provides valid data
- we send valid format list to the destination desktop
- everybody uses valid and correct formats throughout the exchange
- client requests format data _before_ the destination desktop asks it
- client provides malformed data to the destination desktop

Thus I suspect that the following happens:

1. Some third-party application, most likely a clipboard manager or the text editor, requests our clipboard as text.
2. FreeRDP converts *CF_UNICODETEXT* (null-terminated text in UTF-16 with CR-LF lines) to *UTF8_STRING* (text in UTF-8 with LF line without a terminating null byte).
3. FreeRDP provides the conversion result to requesting application, and caches it.
4. Destination desktop requests CF_UNICODETEXT.
5. Client(2) asks Client(1) for CF_UNICODETEXT via raw transfer.
6. Client(1) immediately provides the cached data (which has a wrong format UTF8_STRING).
7. Client(2) takes the data and sends it to the server as is.

With that guess, I am finally able to reproduce the issue with the following steps in my environment:

1. Copy text from desktop 1.
2. Alt-Tab into some local text editor and paste text.
3. Alt-Tab to desktop 2 and paste the text there.

So the culprit should be incorrect cooperation of clipboard data caching and raw transfers. We cache _converted_ data, but for raw transfers we need to cache and provide the original data.

**Comment by FragsterAt** [comment]
I did the same actions in the IceWM, to eliminate the influence of KDE desktop applets, the result was the same.

**Comment by meersjo** [comment]
@ilammy > Unfortunately, the environment I have at the moment does not support remote apps, so I am not able to

Hm, side question: Is this on RAPPS (RAIL), and if so, does that mean there are builds where RAIL clipboard is functional? There's a couple of long-standing issues about that :-)

**Comment by ilammy** [comment]
@meersjo, yeah, I initially thought that this can be something specific to remote apps (RAIL, seamless, whatyoucallit). Then I recalled that clipboard does not work with RAIL in mainline FreeRDP.

I believe this could be fixed by 1) removing `if (!xfc->remote_app)` from xf_event_process(), 2) creating an invisible window to handle clipboard on behalf of the whole remote session when RAIL is active. We've been doing something like this in a closed-source RDP client based on FreeRDP, and it worked well, as I recall.

I guess I could do this as soon as I get myself some Windows VM that actually supports RAIL.

**Comment by ilammy** [comment]
Hi @FragsterAt,

I could not reproduce the issue anymore with the fix from PR #3472. Could you please verify this in your environment? You can get the source code with the fix like this:

```
git clone --branch=x11-cliprdr/raw-data-cache --depth=1 https://github.com/ilammy/FreeRDP.git
```

Thank you.

**Comment by FragsterAt** [comment]
It works!
Now I am waiting for the appearance of patches in a repository opensuse:x11 https://build.opensuse.org/package/show?project=X11%3ARemoteDesktop&package=freerdp via https://ci.freerdp.com/job/freerdp-nightly-rpm-build/.
It is also very important to me correction of https://github.com/FreeRDP/FreeRDP/issues/3439

**Comment by meersjo** [comment]
@ilammy , You may well already know this, but I used a variant of the instructions on http://www.windowsnetworking.com/articles-tutorials/netgeneral/host-remoteapps-windows-pcs-part1.html to enable rapps support on a regular windows desktop :-)

**Comment by tigerfoot** [comment]
@FragsterAt I'm tracking the issue at https://bugzilla.opensuse.org/show_bug.cgi?id=1004108
I'm traveling the rest of the week, so I can't promise an update immediately.

**Comment by akallabeth** [comment]
RAW clipboard data support #3472 merged.
Assuming fixed, closing. Please reopen if still an issue.
